### ISSUE TITLE
Consider type classes for smashing hlists and coproducts

### ISSUE THREAD

**Issue by travisbrown** [milessabin/shapeless#413]
First for some motivation: in the [Finch](https://github.com/finagle/finch) project we're using Shapeless in a couple of places, including a `Router[A]` type that may match a path and return an `A`. We provide an `:+:` method that allows the user to compose routers of different types. This makes combining endpoints into something we can serve a lot cleaner, since we don't have to take every router all the way to a Finagle service—we just need something that we know how to turn into a Finagle service.

For example, we could write something like this:

```scala
val router =
  Get / "foo" / long   :+:
  Get / "foo" / string :+: 
  Get / "baz" / string /> Baz(_)
```

Now we have a `Router[Long :+: String :+: Baz :+: CNil]`, and assuming we have the appropriate type class instances for all the types involved, we can turn this into a web service with `router.toService`.

Finch also allows users to combine coproduct routers with `:+:` in such a way that coproducts are flattened. For example, we might have the following:

```scala
val userEndpoints: Router[User :+: List[User] :+: Service[HttpRequest, User] :+: CNil] =
  Get  / "user" / long /> getUser(_) :+:
  Get  / "user" /> getUsers          :+:
  Post / "user" /> createUser

val ticketEndpoints: Router[Ticket :+: HttpResponse :+: CNil] =
  Get    / "ticket" / long /> getTicket(_) :+:
  Delete / "ticket" / long /> deleteTicket(_)
```

We can serve these routers together like this:

```scala
Await.ready(Httpx.serve(":8080", (userEndpoints :+: ticketEndpoints).toService))
```

This works because `:+:` just smashes the coproducts together into a big coproduct where all the elements are types we know how to serve. This would be a lot more complicated if `userEndpoints :+: ticketEndpoints` were a coproduct of coproducts.

We _could_ introduce a separate operator for this kind of concatenation, but the idea is that we want to keep the API simple. In general Finch users aren't writing Shapeless types in their code—they just put together alternatives with our `:+:` and then turn the whole thing into a service.

This is all supported by a [`CoproductSmash[A, B]` type class](https://github.com/travisbrown/finch/blob/topic/coproduct-smash/core/src/main/scala/io/finch/route/CoproductSmash.scala) that concatenates `A` and `B` if they're both coproducts, prepends `A` to `B` if only `B` is a coproduct, appends `B` to `A` if only `A` is a coproduct, and returns `A :+: B :+: CNil` if neither is a coproduct.

We also have a [`Smash[A, B]` type class](https://github.com/finagle/finch/blob/0.7.0/core/src/main/scala/io/finch/route/Smash.scala) for `HList` that we use in our `RequestReader` type that works similarly:

```scala
smash(1, 'a)                 == 1 :: 'a :: HNil
smash(1, 'a :: HNil)         == 1 :: 'a :: HNil
smash(1 :: HNil, 'a)         == 1 :: 'a :: HNil
smash(1 :: HNil, 'a :: HNil) == 1 :: 'a :: HNil
```

And so on.

These type classes are extremely generic, and I've written them a couple of times for different projects. Have other people written stuff like this? Would there be a place for these in Shapeless so I can stop maintaining them in multiple projects?

**Comment by milessabin** [comment]
Yes, something like this would be useful to have.

It'd be nice to have as much symmetry between the product and coproduct cases as possible here ... in shapeless 3.0 they'll become the same, and it makes sense to anticipate that now. For instance, the value level function makes sense for both products and coproducts if you think of the argument as being a pair (product) vs. an `Either` (coproduct).

I'm not super keen on the name "smash" though. It's cute, and @PLT_Hulk would approve, but I think something more "concat" like would be better.

**Comment by milessabin** [comment]
Fixed in #414.
